staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -121,6 +132,197 @@ const (
 // +genclient:nonNamespaced
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// ValidatingAdmissionPolicy describes the definition of an admission validation policy that accepts or rejects and object without changing it.
+type ValidatingAdmissionPolicy struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata; More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata.
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Specification of the desired behavior of the AdmissionPolicy.
+	// +optional
+	Spec AdmissionPolicySpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ValidatingAdmissionPolicyList is a list of ValidatingAdmissionPolicy.
+type ValidatingAdmissionPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// List of ValidatingAdmissionPolicy.
+	Items []ValidatingAdmissionPolicy `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// AdmissionPolicySpec is the specification of the desired behavior of the AdmissionPolicy.
+type AdmissionPolicySpec struct {
+	//ParamSource specifies the kind of resources used to parameterize this policy
+	// +optional
+	ParamSource ParamSource `json:"ParamSource,omitempty" protobuf:"bytes,1,rep,name=ParamSource"`
+
+	// MatchConstraints specifies what resources this policy is designed to validate.
+	// The AdmissionPolicy cares about an validation if it matches _any_ Constriant.
+	// However, in order to prevent ValidatingAdmissionPolicy
+	// from putting the cluster in a state which cannot be recovered from without completely
+	// disabling the process, ValidatingAdmissionPolicy is never called
+	// on admission requests for AdmissionPolicySpec objects.
+	// +optional
+	MatchConstraints []RuleWithOperations `json:"MatchConstraints,omitempty" protobuf:"bytes,2,rep,name=matchConstraints"`
```

Comment:
cc @maxsmythe here too

Reply:
IIRC there was some talk about `kind` being useful to identify the schema (apparently it's more specific than "resource" in identifying the actual schema used?)

Would there be any other reasons for the selectors to diverge?

If not, and it's purely for allowing type safety where applicable, it may be worth creating a specific entry just for typing, since other features would need to be disallowed, such as `"*"` to guarantee a bounded set of types.

Then again, type safety is a heavy lift, I'm not sure if we know enough about what's possible to know what the schema would require. E.g., is it possible to check for the validity of fields accessed across all cited types, or would we need separate CEL per type?

TL;DR type safety is likely different enough that we probably don't want the field providing that doing dual duty with other types of selectors, so curious about other use cases for divergence.

Reply:
I have renamed to use MatchResources in both places. Please let me know if other way is preferred.